packages/css-engine/src/core/style-sheet.ts

```diff
@@ -1,15 +1,101 @@
-import type { StyleElement } from "./style-element";
+import type { Style } from "../schema";
+import {
+  FontFaceRule,
+  MediaRule,
+  PlaintextRule,
+  type FontFaceOptions,
+  type MediaRuleOptions,
+} from "./rules";
+import { compareMedia } from "./compare-media";
+import { StyleElement } from "./style-element";
+
+export type CssRule = {
+  style: Style;
+  breakpoint?: string;
+};
 
 export class StyleSheet {
   #cssText = "";
-  #element;
+  #mediaRules: Map<string, MediaRule> = new Map();
+  #plainRules: Map<string, PlaintextRule> = new Map();
+  #fontFaceRules: Array<FontFaceRule> = [];
+  #isDirty = false;
+  #element: StyleElement;
   constructor(element: StyleElement) {
     this.#element = element;
   }
-  replaceSync(cssText: string) {
-    if (cssText !== this.#cssText) {
-      this.#cssText = cssText;
-      this.#element.render(cssText);
+  addMediaRule(id: string, options?: MediaRuleOptions) {
+    let mediaRule = this.#mediaRules.get(id);
+    if (mediaRule === undefined) {
+      mediaRule = new MediaRule(options);
+      this.#mediaRules.set(id, mediaRule);
+      this.#isDirty = true;
+      return mediaRule;
+    }
+
+    if (options) {
+      mediaRule.options = options;
+      this.#isDirty = true;
+    }
+
+    return mediaRule;
+  }
+  addPlaintextRule(cssText: string) {
+    const rule = this.#plainRules.get(cssText);
+    if (rule !== undefined) {
+      return rule;
+    }
+    this.#isDirty = true;
+    return this.#plainRules.set(cssText, new PlaintextRule(cssText));
+  }
+  addFontFaceRule(options: FontFaceOptions) {
+    this.#isDirty = true;
+    return this.#fontFaceRules.push(new FontFaceRule(options));
+  }
+  markAsDirty() {
+    this.#isDirty = true;
+  }
+  get cssText() {
+    if (this.#isDirty === false) {
+      return this.#cssText;
     }
+    this.#isDirty = false;
+    const css: Array<string> = [];
+
+    css.push(...this.#fontFaceRules.map((rule) => rule.cssText));
```

Comment:
Am I right we want to inline this into html in the near future?

Reply:
I don't know it makes sense to completely stop caching css files, we need to do more benchmarking, I guess it would make sense for smaller sites that change very often, and make less sense for larger sites like our landing that changes less often

- The more it changes the less there is benefit in caching - so inlining makes more sense with frequent changes
- The bigger the file the less it makes sense to inline

Reply:
I mean #fontFaceRules

Reply:
oh yeah, but you can use the engine, then get cssText and inline it